From 256b6a3decc150b4663dff3f2e47885dc36638ad Mon Sep 17 00:00:00 2001
From: Kieran Bingham <kieran.bingham@ideasonboard.com>
Date: Sun, 21 Jun 2026 00:00:28 +0100
Subject: [PATCH] libcamera: vector: Add clamp operation

Provide a helper that will perform a std::clamp given a scaler high and
low value to clamp to.

Suggested-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Kieran Bingham <kieran.bingham@ideasonboard.com>
---
 include/libcamera/internal/vector.h | 8 ++++++++
 src/libcamera/vector.cpp            | 8 ++++++++
 test/vector.cpp                     | 1 +
 3 files changed, 17 insertions(+)

diff --git a/include/libcamera/internal/vector.h b/include/libcamera/internal/vector.h
index ed7490e..5f7d450 100644
--- a/include/libcamera/internal/vector.h
+++ b/include/libcamera/internal/vector.h
@@ -171,6 +171,14 @@ public:
 		return apply(*this, scalar, [](T a, T b) -> T { return std::max(a, b); });
 	}
 
+	constexpr Vector clamp(T low, T high) const
+	{
+		Vector result;
+		for (unsigned int i = 0; i < Rows; i++)
+			result[i] = std::clamp(data_[i], low, high);
+		return result;
+	}
+
 	constexpr T dot(const Vector<T, Rows> &other) const
 	{
 		T ret = 0;
diff --git a/src/libcamera/vector.cpp b/src/libcamera/vector.cpp
index 86b9f9b..88c1959 100644
--- a/src/libcamera/vector.cpp
+++ b/src/libcamera/vector.cpp
@@ -210,6 +210,14 @@ LOG_DEFINE_CATEGORY(Vector)
  * \return The element-wise maximum of this vector and \a scalar
  */
 
+/**
+ * \fn Vector::clamp(T low, T high) const
+ * \brief Clamp the vector element-wise between \a low and \a high
+ * \param[in] low The lower limit
+ * \param[in] high The upper limit
+ * \return A vector with each element clamped between \a low and \a high
+ */
+
 /**
  * \fn Vector::dot(const Vector<T, Rows> &other) const
  * \brief Compute the dot product
diff --git a/test/vector.cpp b/test/vector.cpp
index 4fae960..2f5d851 100644
--- a/test/vector.cpp
+++ b/test/vector.cpp
@@ -72,6 +72,7 @@ protected:
 		ASSERT_EQ(v2.min(4.0), (Vector<double, 3>{{ 1.0, 4.0, 4.0 }}));
 		ASSERT_EQ(v2.max(v3), (Vector<double, 3>{{ 4.0, 4.0, 8.0 }}));
 		ASSERT_EQ(v2.max(4.0), (Vector<double, 3>{{ 4.0, 4.0, 8.0 }}));
+		ASSERT_EQ(v2.clamp(2.0, 6.0), (Vector<double, 3>{{ 2.0, 4.0, 6.0 }}));
 
 		ASSERT_EQ(v2.dot(v3), 52.0);
 
-- 
2.43.0

